### PR TITLE
PWN-4272 - New Pin code flow

### PR DESCRIPTION
**Pull request by kuznetsov**
### Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-4272

### Description of Work

Redesigned Pin code screen and added new Touch ID screen

### Screenshots

Screenshot 2022-07-28 at 10 09 53 | Screenshot 2022-07-28 at 10 10 13
:------------:|:-------------:|

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/auth/AuthModule.kt [hunk 1]
```diff
@@ -73,5 +77,7 @@ object AuthModule {
         factory { WalletWeb3AuthManager(get(), get(), get(), get()) }
 
         factory { NewOnboardingPresenter(get()) } bind NewOnboardingContract.Presenter::class
+        factory { NewCreatePinPresenter(get(), get()) } bind NewCreatePinContract.Presenter::class
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/AuthModule.kt:80]: `factoryOf`

File: app/src/main/java/org/p2p/wallet/auth/ui/pin/newcreate/NewCreatePinFragment.kt [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    override val presenter: NewCreatePinContract.Presenter by inject()
+
+    private val binding: FragmentNewCreatePinBinding by viewBinding()
+    private val analyticsInteractor: ScreensAnalyticsInteractor by inject()
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        with(binding) {
+            toolbar.setNavigationOnClickListener { presenter.clearUserData() }
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/ui/pin/newcreate/NewCreatePinFragment.kt:37]: should we navigate back?

File: app/src/main/java/org/p2p/wallet/auth/ui/pin/newcreate/NewCreatePinFragment.kt [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    override fun showLoading(isLoading: Boolean) {
+        binding.pinView.showLoading(isLoading)
+    }
+
+    override fun showCreation() {
+        with(binding) {
+            pinView.isEnabled = true
+            textViewTitle.text = getString(R.string.auth_create_wallet_set_up_your_pin)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/ui/pin/newcreate/NewCreatePinFragment.kt:59]:
```suggestion
            textViewTitle.setText(R.string.auth_create_wallet_set_up_your_pin)
```

File: app/src/main/java/org/p2p/wallet/auth/ui/pin/select/TouchIdContract.kt [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package org.p2p.wallet.auth.ui.pin.select
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/ui/pin/select/TouchIdContract.kt:1]: what `select` package here means?

File: app/src/main/res/layout/fragment_new_create_pin.xml [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            android:id="@+id/toolbar"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            app:layout_constraintTop_toTopOf="parent"
+            app:navigationIcon="@drawable/ic_back_night"
+            app:title="@string/auth_create_wallet_step_3"
+            app:titleTextAppearance="@style/WalletTheme.TextAppearance.SemiBold18" />
+
+        <ImageView
+            android:id="@+id/imageViewLock"
```
Comment (gslevinkov) [inline app/src/main/res/layout/fragment_new_create_pin.xml:25]: we haven't decided about changing the `type` place in the naming
types should still be at the end of the name

File: ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitNumberKeyboardButtonView.kt [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import org.p2p.uikit.R
+import org.p2p.uikit.databinding.WidgetUiKitNumberButtonBinding
+
+class UiKitNumberKeyboardButtonView @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyleAttr: Int = 0
+) : FrameLayout(context, attrs, defStyleAttr) {
+
+    private val binding = WidgetUiKitNumberButtonBinding.inflate(LayoutInflater.from(context), this)
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitNumberKeyboardButtonView.kt:21]: we have inflateViewBinding ext now

File: ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitNumberKeyboardButtonView.kt [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+    init {
+        attrs?.let {
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitNumberKeyboardButtonView.kt:24]: we haven't checked attrs for nullability in other custom views

File: ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitNumberKeyboardButtonView.kt [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+                binding.keyboardButtonTextView.text = text
+            }
+
+            val image = typedArray.getDrawable(R.styleable.UiKitNumberKeyboardButtonView_keyboard_button_image)
+            if (image != null) {
+                binding.keyboardButtonImageView.setImageDrawable(image)
+                binding.keyboardButtonImageView.visibility = View.VISIBLE
+            }
+
+            typedArray.recycle()
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitNumberKeyboardButtonView.kt:38]: use?

File: ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitNumberKeyboardButtonView.kt [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+        setBackgroundResource(R.drawable.bg_new_keyboard_selector)
+
+        clipToOutline = true
+    }
+
+    fun setIcon(@DrawableRes drawableResId: Int) {
+        binding.keyboardButtonImageView.setImageResource(drawableResId)
+        binding.keyboardButtonImageView.visibility = View.VISIBLE
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitNumberKeyboardButtonView.kt:48]: isVisible

File: ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitNumberKeyboardButtonView.kt [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    fun setPressedColor(color: Int) {
+        val drawable = ContextCompat.getDrawable(context, R.drawable.bg_new_keyboard_pressed) as RippleDrawable
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitNumberKeyboardButtonView.kt:52]: binding.getDrawable

File: ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitNumberKeyboardView.kt [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+                    R.id.onePinCodeButton -> '1'
+                    R.id.twoPinCodeButton -> '2'
+                    R.id.threePinCodeButton -> '3'
+                    R.id.fourPinCodeButton -> '4'
+                    R.id.fivePinCodeButton -> '5'
+                    R.id.sixPinCodeButton -> '6'
+                    R.id.sevenPinCodeButton -> '7'
+                    R.id.eightPinCodeButton -> '8'
+                    R.id.ninePinCodeButton -> '9'
+                    else -> throw IllegalArgumentException("ID does not reference a View inside this View")
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitNumberKeyboardView.kt:70]: you can put view.id in the throwable too

File: ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitPinCodeView.kt [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+    private var roundViews = mutableListOf<ImageView>()
+    private var emptyDotDrawableId: Drawable? = null
+    private var fullDotDrawableId: Drawable? = null
+
+    var currentLength: Int = 0
+        private set
+
+    init {
+        attrs.let {
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitPinCodeView.kt:33]: no need to check for attrs here IMHO

File: ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitPinCodeView.kt [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+            } else {
+                roundViews[i].setImageDrawable(emptyDotDrawableId)
+            }
+        }
+    }
+
+    @SuppressWarnings("MagicNumber")
+    fun startErrorAnimation(onAnimationFinished: () -> Unit) {
+        val animation = TranslateAnimation(-4f, 4f, 0f, 0f)
+        animation.duration = ANIMATION_DURATION
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitPinCodeView.kt:54]: apply?

File: ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitPinCodeView.kt [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+                roundViews[i].setImageDrawable(fullDotDrawableId)
+            } else {
+                roundViews[i].setImageDrawable(emptyDotDrawableId)
+            }
+        }
+    }
+
+    @SuppressWarnings("MagicNumber")
+    fun startErrorAnimation(onAnimationFinished: () -> Unit) {
+        val animation = TranslateAnimation(-4f, 4f, 0f, 0f)
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitPinCodeView.kt:53]: magic numbers

File: ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitPinCodeView.kt [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+        animation.duration = ANIMATION_DURATION
+        animation.repeatMode = Animation.REVERSE
+        animation.repeatCount = 2
+        animation.setAnimationListener(object : Animation.AnimationListener {
+            override fun onAnimationEnd(animation: Animation?) {
+                setDotsColor(null)
+                onAnimationFinished()
+            }
+
+            override fun onAnimationRepeat(animation: Animation?) {}
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitPinCodeView.kt:63]:
```suggestion
            override fun onAnimationRepeat(animation: Animation?) = Unit
```

File: ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitPinCodeView.kt [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+            override fun onAnimationStart(animation: Animation?) {
+                setDotsColor(R.color.mint)
+            }
+
+            override fun onAnimationEnd(animation: Animation?) {
+                setDotsColor(null)
+                onAnimationFinished()
+            }
+
+            override fun onAnimationRepeat(animation: Animation?) {}
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/organisms/pin/UiKitPinCodeView.kt:87]:
```suggestion
            override fun onAnimationRepeat(animation: Animation?) = Unit
```

File: app/src/main/java/org/p2p/wallet/auth/ui/pin/select/TouchIdContract.kt (hunk 4 above)
Comment (kuznetsov) [reply app/src/main/java/org/p2p/wallet/auth/ui/pin/select/TouchIdContract.kt:1]: It was temporary, will be renamed

File: app/src/main/java/org/p2p/wallet/auth/ui/pin/newcreate/NewCreatePinFragment.kt (hunk 2 above)
Comment (kuznetsov) [reply app/src/main/java/org/p2p/wallet/auth/ui/pin/newcreate/NewCreatePinFragment.kt:37]: `navigateBack()` is called from presenter in that case

Comment (kuznetsov) [reply app/src/main/java/org/p2p/wallet/auth/ui/pin/newcreate/NewCreatePinFragment.kt]: But clearing user data is not needed here, so I just navigate back

**Review by gslevinkov — APPROVED** [review]

**Pull request merged**